src/encoding.ts

```diff
@@ -0,0 +1,379 @@
+		}
+
+		let decoded = '';
+
+		for (let i = 0, length = data.length; i < length; i++) {
+			let byte1 = data[i] & 0xFF;
+
+			if ((byte1 & 0x80) === 0) {
+				decoded += decodeUtf8EncodedCodePoint(byte1);
+			}
+			else if ((byte1 & 0xE0) === 0xC0) {
+				let byte2 = data[++i] & 0xFF;
+				validateUtf8EncodedCodePoint(byte2);
+				byte2 = byte2 & 0x3F;
+				let encodedByte = ((byte1 & 0x1F) << 0x06) | byte2;
+				decoded += decodeUtf8EncodedCodePoint(encodedByte, [0x80, Infinity]);
+			}
+			else if ((byte1 & 0xF0) === 0xE0) {
+				let byte2 = data[++i] & 0xFF;
+				validateUtf8EncodedCodePoint(byte2);
+				byte2 = byte2 & 0x3F;
+
+				let byte3 = data[++i] & 0xFF;
+				validateUtf8EncodedCodePoint(byte3);
+				byte3 = byte3 & 0x3F;
+
+				let encodedByte = ((byte1 & 0x1F) << 0x0C) | (byte2 << 0x06) | byte3;
+				decoded += decodeUtf8EncodedCodePoint(encodedByte, [0x0800, Infinity], true);
+			}
+			else if ((byte1 & 0xF8) === 0xF0) {
+				let byte2 = data[++i] & 0xFF;
+				validateUtf8EncodedCodePoint(byte2);
+				byte2 = byte2 & 0x3F;
+
+				let byte3 = data[++i] & 0xFF;
+				validateUtf8EncodedCodePoint(byte3);
+				byte3 = byte3 & 0x3F;
+
+				let byte4 = data[++i] & 0xFF;
+				validateUtf8EncodedCodePoint(byte4);
+				byte4 = byte4 & 0x3F;
+
+				let encodedByte = ((byte1 & 0x1F) << 0x0C) | (byte2 << 0x0C) | (byte3 << 0x06) | byte4;
+				decoded += decodeUtf8EncodedCodePoint(encodedByte, [0x010000, 0x10FFFF]);
+			}
+			else {
+				validateUtf8EncodedCodePoint(byte1);
+			}
+		}
+
+		return decoded;
+	}
+}
+
+/**
+ * Hex codec that provides facilities for encoding a string (or string coerced object)
+ * into an Hex encoded byte buffer as well as decoding an Hex encoded byte buffer
+ * into a string.
+ */
+export class Hex {
+	/**
+	 * Encode a string or string coerced object into an Hex byte buffer
+	 *
+	 * @param data
+	 * Any object that is either a string or can be coerced into a string
+	 */
+	static encode(data: any): ByteBuffer {
+		let buffer: ByteBuffer;
+
+		[ data, buffer ] = normalizeEncodingArgs(data, Hex);
+
+		for (let i = 0, length = data.length; i < length; i++) {
+			let encodedChar = parseInt(data.charCodeAt(i).toString(16), 16);
+
+			buffer[i] = encodedChar;
+		}
+
+		return buffer;
+	}
+
+	/**
+	 * Decode a Hex encoded byte buffer into a string
+	 *
+	 * @param data
+	 * The byte buffer to be decoded
+	 */
+	static decode(data: ByteBuffer): string {
+		if (!validateDecodingArgs(data)) {
+			return '';
+		}
+
+		let decoded = '';
+
+		for (let i = 0, length = data.length; i < length; i++) {
+			let decodedChar = String.fromCharCode(data[i]);
+
+			decoded += decodedChar;
+		}
+
+		return decoded;
+	}
+}
+
+/**
+ * Base64 codec that provides facilities for encoding a string (or string coerced object)
+ * into an Base64 encoded byte buffer as well as decoding an Base64 encoded byte buffer
+ * into a string.
+ * Adapted from http://www.nczonline.net/blog/2009/12/08/computer-science-in-javascript-base64-encoding/
```

Comment:
This should actually be able to use `atob` and `btoa` in all but IE9 (possibly with [some extra considerations for Unicode](https://developer.mozilla.org/en-US/docs/Web/API/WindowBase64/btoa#Unicode_Strings)).

Meanwhile, Dojo 1 already has base64 code that we can lift from (though the code is unfortunately harder to read...)